File: 414-ThirdMaximumNumber/414-ThirdMaximumNumber.py
# Last updated: 12/28/2025, 4:53:10 PM
class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        l=set(nums)
        nums=list(l)
        # nums.sort()
        if len(nums)<=2:
            return max(nums)
        else:
            for i in range(0,2):
                nums.remove(max(nums))
            return max(nums)

File: 414-ThirdMaximumNumber/414-ThirdMaximumNumber.java
// Last updated: 12/28/2025, 4:53:07 PM
class Solution {
    public int thirdMax(int[] nums) {
        Set<Integer> numSet = new HashSet<>();
        for(int num:nums){
            numSet.add(num);
        } 
        int[] uniqueNums = new int[numSet.size()];
        int index=0;
        for(int num:numSet){
            uniqueNums[index++]=num;
        }
        if(numSet.size()<3){
            int max=Integer.MIN_VALUE;
            for(int num: numSet){
                max=Math.max(max,num);
            }
            return max;
        }
        
        PriorityQueue<Integer>minHeap = new PriorityQueue<>();
        for(int num: uniqueNums){
            minHeap.add(num);
            if(minHeap.size()>3){
                minHeap.poll();
            }
        }
        int ans=minHeap.poll();
        return ans;
    }
}